pkg/worker/worker.go

```diff
@@ -0,0 +1,244 @@
+and provide a correct DataStorage. Control the processes using Start() and Stop()
+methods. Query the output anytime, even when the process is running using Query()
+method.
+
+DataStorage interface must include working data synchronization methods: Lock()
+and Unlock().
+*/
+package worker
+
+import (
+	"bufio"
+	"log"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+
+	googleUUID "github.com/google/uuid"
+)
+
+var (
+	// ErrJobNotFound indicates that the requested job
+	// has not been found in the DataStorage.
+	ErrJobNotFound = errors.New("job has not been found")
+
+	// ErrJobAlreadyStopped indicates that the requested job
+	// has already been stopped.
+	ErrJobAlreadyStopped = errors.New("job has already been stopped")
+)
+
+// DataStorage allows to store and retrieve jobs from any storage that implements the interface.
+// The storage needs to implement sync.Mutex or other form of access synchronization based on
+// Lock() and Unlock() methods.
+type DataStorage interface {
+	Store(job Job) error
+	Retrieve(uuid googleUUID.UUID) (Job, error)
+	sync.Locker
+}
+
+// Worker provides an interface for running, stopping and querying processes.
+type Worker struct {
+	DataStorage DataStorage
+}
+
+// NewWorker constructs a new Worker with a struct that implements DataStorage interface.
+func NewWorker(dataStorage DataStorage) *Worker {
+	return &Worker{DataStorage: dataStorage}
+}
+
+// Start creates a new Linux process with the provided command,
+// instantiates a new Job object and returns its ID.
+// It runs the job in a goroutine that updates its progress directly
+// to the DataStorage.
+// TODO: Add a proper logger (e.g. logrus) interface to the Worker struct to standarize logging.
+// TODO: Report errors and job completion status via channels.
+// TODO: Add a optional timeout for better control over goroutines.
+//
+// Notes: I generally try to avoid using goroutines in packages. It introduces unpredictable
+// side effects and takes control away from the client. I couldn't come up with a better
+// solution to allow this package to control StdIn and StdErr of the process.
+func (worker *Worker) Start(command string, user string) googleUUID.UUID {
+	job := newJob(worker, command, user)
+
+	go func() {
+		err := job.start()
+		if err != nil {
+			log.Printf("the following errors arised when running job %v: %s", job.UUID, err)
+		}
+	}()
+
+	return job.UUID
+}
+
+// Stop stops the job by its UUID and stores the timestamp.
+// It returns named errors when the requested job is either not found or has already
+// been stopped.
+func (worker *Worker) Stop(uuid googleUUID.UUID) error {
+	worker.DataStorage.Lock()
+	job, err := worker.DataStorage.Retrieve(uuid)
+	worker.DataStorage.Unlock()
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving job with UUID: %v", uuid)
+	}
+	if job.UUID == googleUUID.Nil {
+		return ErrJobNotFound
+	}
```

Comment:
It would be a bit cleaner to have `DataStoage.Retrieve` return a pointer to a job.
Or have it return `ErrJobNotFound`, which can also be easily checked.

Reply:
Actually, it's what I had in mind (returning ErrJobNotFound) but somehow missed it. I'll fix it.